sdnfv/openNetVM
# Shared core functionality for messages

Adds functionality to sleep when no messages are enqueued onto an nf's message ring. 


## Summary:
This is functionality I implemented as part of the larger mTCP project, in which an NF constantly receives lots of messages. This works exactly the same way as sleeping on an empty packet ring. 

**Usage:**
Run the manager with shared CPU enabled.


| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        | X
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 



## Merging notes:
 - Dependencies: None  

**TODO before merging :**



## Test Plan:
Tested by sending a large influx of messages to an NF running in shared cpu mode. Also tested with multiple NF's on the same core. Checked htop CPU usage to verify proper sleeping when no messages were being sent. 


## Review: 
### Review checklist:  

- Sanity checks, assigned to @koolzz  @kevindweb 
    - Run make in `/onvm` and `/examples` directories
    - Test the new functionality

- Code style, assigned to @koolzz  @kevindweb
    - Run linter
    - Check everything style related

- Code design, assigned to @koolzz  @kevindweb
    - Check for memory leaks
    - Check that code is reusable
    - Code is clean, functions are concise
    - Verify that edge cases properly handled

- Performance, assigned to @koolzz  @kevindweb
    - Run Speed Tester NF, report performance.
    - Run pktgen, report performance (not really required)

- Documentation, assigned to @koolzz  @kevindweb
    - Check if the new changes are well documented, in both code and READMEs
